---
title: "Data Structures in R"
author: "Fred LaPolla"
date: "7/13/2020"
output: slidy_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo = TRUE}
install.packages("RCurl", repos = "http://cran.us.r-project.org")
library(RCurl)
url <- getURL("https://raw.githubusercontent.com/fredwillie/R_PhDProg_NYU_2020/master/Crusio1.csv")
crusio1 <- read.csv(text = url)
```


## Data Structures in R

</br>
</br>
</br>

R holds several structures of data. In order from most complex to most simple:

>- Dataframes
>- Matrices
>- Lists
>- Vectors

***

## Dataframes
</br>
</br>
</br>
**Note:** My background is such that spreadsheets (like Excel) are a normal way to store/organize data.

A dataframe is basically R's spreadsheet. 

***

## Dataframes

</br>
</br>
</br>

Dataframes have columns and rows of data. The data in these different columns can be of different types. Imagine making a spreadsheet of people in your lab: their name (character), their age (integer), their field of study (factor). Each column will have one type of data, but any given column can be a different data type from the column next to it. 

Each row contains all the types of data for any given record, or in this case, member of your lab. 

***

## Matrices

</br>
</br>
</br>

A matrix is like a dataframe but all data in the columns and rows are of the **same type**.

This may sound trivial but is important to bioinformatics work!

If we have a matrix of gene expression data and want to make a heatmap, every cell must contain a number. If there is a stray mark or we read in data, we may need to convert the data after. Even one letter in a cell will lead R to conclude that all of the numbers are actually characters. 



***

## Lists

</br>
</br> 
</br>

Lists are like one row of a spreadsheet. Using our example from above, a list might contain the name of one person in your lab, their age, their field of work and other information. 

***

## Vectors

</br>
</br>
</br>

A vector is like a list but with everything of one type of data: i.e. all numbers, all factors, etc. 

If you perform some operation on vectors, the vector will be processed one after the next. If you perform an operation on two vectors, for example multiplying them, the operations will go down the list of each vector:

```{r }
x <- c(1,2,3,4)
y <- c(2,3)
x*y
```

***

## Binding Vectors

</br>
</br>
</br>

You can combine vectors to be a matrix with rbind() for combining rows and cbind() for columns. If the columns or rows are evenly divisible, they will recycle, but if for example you tried to r bind a vector with 4 objects and a vector with 5 objects, you will get an error. 

```{r}
rbind(x, y)

z <- c(5,8,0,9,10)
rbind(x,z)
```


You can also apply names using names() for example:

```{r}
names(x)<- c("First", "Second", "Third", "Fourth")
x
```


***

## Finding out what kind of data structure

</br>
</br>
</br>

Just like data type, you can use class on a whole data object to find out what structure it has:

```{r}
class(crusio1)
```



***

## Let's explore some data

Find a dataset and read it in. If you do not have one, an easy option might be last week's Crusio1 data.

*** 

## How to tell what kind of data you are working with

What kind of data is in your data set?

```{r}


class(crusio1)

class(crusio1$strain)

class(crusio1$sex)

class(crusio1$attack_combine)

class(crusio1$bw)

class(crusio1[4,10])
```


This can be applied to an object like a dataframe or matrix, or to individual pieces of data


*** 

## Exploration

What kind of data are you working with?


***

## Changing data types

Frequently when we load data into R, R will assume the wrong type of data. For example in our crusio1, look at what type of data attack day 2 is classified as:

```{r}
class(crusio1$attack_d2)
```

Integer feels wrong for a binary situation (like a grouping), so we can reset it using a factor command. Let's look up what the factor command offers:

```{r}
?factor
```

The main things to note are x, the vector we are transforming (e.g. a column in a dataframe), levels (i.e. how many groups are there), labels (i.e. what are the codes for in text)

In the code below we are going to assign groups, or levels, to this factor and overwrite the existing column with this data. *WARNING* Only run this command once, as once we overwrite the command, running it again will yield all NAs and we will need to pull in the data to correct it. 

```{r}
crusio1$attack_d2 <- factor(crusio1$attack_d2, levels = 0:1, labels = c("No_Attack", "Attack"))
summary(crusio1$attack_d2)
```

*** 

## Confusing Note: as.factor() and factor(), as.matrix() vs matrix() etc.

</br>
</br>
</br>

You can often coerce data types into another type with as.[data type]. For example:

```{r}
as.integer(TRUE)
```


Normally, the differences between for example as.factor() or factor() or as.matrix() or matrix() are not always super apparent. If your results are not what you expect, you might try as.factor() vs factor() to see if one works how you want.



*** 

## Reclassifying Data Structures

As we have seen, sometimes R reads in data as the wrong type, but it can also read in the wrong data structure. Imagine we had a matrix that when we pulled in R thought it was a dataframe:

```{r}
WannaMatrix <- crusio1[5:10, 18:22]
head(WannaMatrix)
class(WannaMatrix)
```

*** 

## Reclassifying Data Structures


Here we can reclassify it to be a matrix:

```{r}
WannaMatrix <- as.matrix(WannaMatrix,nrow = 6, ncol = 4) 
class(WannaMatrix)
class(WannaMatrix[4,1])
```

This may become important if you read in data in the wrong format, as visualizations like heatmaps will likely be done in matrices rather than dataframes.  
